Contrast mode (only after user-panel!)

BogdanScherban opened the pull request:
@fzaninotto @tony-shannon 

Contrast mode possibility was added. Results on the video:
https://drive.google.com/file/d/1fDEaWDSJchHk566NB_NtJOSuzUtSP2rf/view

But this PR can be approved ONLY after previous one:
https://github.com/PulseTile/PulseTile-RA/pull/28

Because **contrast-mode** branch is beased on **user-panel** branch.

tony-shannon commented:
thanks

@fzaninotto is the order of accepting this PRs critical? Bogdan has followed your suggestion in terms of basing I think..

BogdanScherban commented:
This PR includes all changes from: https://github.com/PulseTile/PulseTile-RA/pull/28

It is more convenient to review #30 after #28

fzaninotto commented on src/version/features/ContrastMode/index.js:
```diff
@@ -0,0 +1,57 @@
+
+    toggleContrastMode = () => {
+        this.setState(
+            { isContrastMode: !this.state.isContrastMode },
```
again, when you update the state depending on the previous state, use the function form or you risk race conditions

```js
this.setState(state => ({ isContrastMode: !!state.isContrastMode }), ...)
```

BogdanScherban replied:
Fixed

fzaninotto commented on src/version/features/ContrastMode/index.js:
```diff
@@ -0,0 +1,57 @@
+                    className={classes.rightBlockButton}
+                    aria-haspopup="true"
+                    color="inherit"
+                    onClick={() => this.toggleContrastMode()}>
```
`onClick={this.toggleContrastMode}`, or you're creating a new function at each render, and defeating the props identity check.

BogdanScherban replied:
Fixed

fzaninotto commented on src/version/features/ContrastMode/index.js:
```diff
@@ -0,0 +1,57 @@
+class ContrastMode extends Component {
+
+    state = {
+        isContrastMode: this.props.contrastMode,
```
is it in the global state or in the local state? There should be only one source of truth.

BogdanScherban replied:
Changed to: 
state = {
        isContrastMode: false,
}

fzaninotto commented on src/version/sagas/contrastModeSagas.js:
```diff
@@ -0,0 +1,13 @@
+
+import { CONTRAST_MODE_ACTION, contrastModeAction } from "../actions/contrastModeAction";
+
+export default takeEvery(CONTRAST_MODE_ACTION.REQUEST, function*(action) {
```
you don't need a saga for that, it's not a side effect and it's not asynchronous. A reducer is enough and less complex.

BogdanScherban replied:
Saga was removed

fzaninotto commented on src/core/common/CustomLayout.js:
```diff
@@ -50,4 +54,11 @@ const CustomLayout = ({ classes, ...rest }) => {
 
-export default withStyles(styles)(CustomLayout);
+const mapStateToProps = state => {
+    const isContrastMode = get(state, 'custom.contrastMode.data', false);
```
don't use `get` here, it's useless.

BogdanScherban replied:
I can't remove get() here. 

Contrast mode is non-core feature. And this feature can be absent in project. In this case _state.custom.contrastMode.data_ will return error.

Because current file is located in core path. So, **isContrastMode** should be defined correctly in both cases.

fzaninotto replied:
you're right, thanks for the explanation.

fzaninotto commented on src/core/common/CustomLayout.js:
```diff
@@ -50,4 +54,11 @@ const CustomLayout = ({ classes, ...rest }) => {
+const mapStateToProps = state => {
+    const isContrastMode = get(state, 'custom.contrastMode.data', false);
+    return {
+        theme: getCurrentTheme(isContrastMode),
```
warning: `getCurrentTheme` returns a new object each time you call it... this will harm the application performance big time.

BogdanScherban replied:
Yes, it is understandable. But I don't know other way to provide possibility to toggle contrast mode. The main parameter of contrast mode is black color instead of theme color

fzaninotto replied:
use lodash's memoize.

https://lodash.com/docs/4.17.11#memoize

BogdanScherban commented:
@fzaninotto @tony-shannon 

I fixed all notes besides two, where I described a reason.
I updated this PR, but this PR can be merged ONLY after user-panel PR (#28)

fzaninotto commented on src/core/token.js:
```diff
@@ -1 +1,21 @@
+}
+
+export const token = getTokenFromCookie();
+export const domainName = "http://dev.ripple.foundation";
```
this should come from a global config variable

BogdanScherban replied:
@fzaninotto 
Could you please clarify, do you mean React context using or simple globar variables like in JS?

fzaninotto replied:
```
// index.html

<script>
windows.config = {
    mainColor: "#fff"
}
</script>
<script src="/static/main.js" /> 

// in scr/theme.js
const mainColor = windows && windows.config ? global.config.mainColor : '#efe';
```

see https://immutablewebapps.org/

fzaninotto commented on src/version/actions/contrastModeAction.js:
```diff
@@ -4,6 +4,4 @@ export const CONTRAST_MODE_ACTION = createRequestTypes('CONTRAST_MODE_ACTION');
 
 export const contrastModeAction = {
     request: data => ({ type: CONTRAST_MODE_ACTION.REQUEST, data }),
```
IMO, you don't need to export an object since there is only one action. Something like 

```js
export const setContrastMode = data => ({ type: CONTRAST_MODE_ACTION.REQUEST, data })
``` 

Will be enough:

Also, I'm not fan of the generic `data` argument here. It should be clear by reading the function signature what data it expects. By reading at the reducer, I understand that `data` is treated as a boolean, so I'd write the actino as follows:

```js
export const setContrastMode = isEnabled => ({ type: CONTRAST_MODE_ACTION.REQUEST, isEnabled })
```

fzaninotto commented on src/version/common/Topbar/fragments/UserPanelButton.js:
```diff
@@ -35,50 +35,59 @@ const styles = {
+                        className={classes.rightBlockButton}
+                        aria-owns={isOpen ? 'menu-appbar' : undefined}
+                        aria-haspopup="true"
+                        onClick={this.handleMenu.bind(this)}
```
no need for the bind here, as you defined `handleMenu` using the arrow function in the class properties.
